From 69a083fa841bb49ecf8bd97697e1d438315fb7c1 Mon Sep 17 00:00:00 2001
From: Haoran Wang <b50027@freescale.com>
Date: Wed, 23 Dec 2015 17:10:03 +0800
Subject: [PATCH 22/22] MA-7407 Do not disconnect when in WPS process

The bcmdhd will disconnect wps connection when received wps
connection status changed information.

Prevent this notify in supplicant to fix the WPS no
work issue.

Cherry-picked from ca6d9f19.
Since src/drivers/driver_nl80211.c has move to
src/drivers/driver_nl80211_event.c

Signed-off-by: Haoran Wang <b50027@freescale.com>
---
 src/drivers/driver_nl80211_event.c | 8 ++++++++
 wpa_supplicant/Android.mk          | 1 +
 2 files changed, 9 insertions(+)

diff --git a/src/drivers/driver_nl80211_event.c b/src/drivers/driver_nl80211_event.c
index 7b0f721..31de696 100644
--- a/src/drivers/driver_nl80211_event.c
+++ b/src/drivers/driver_nl80211_event.c
@@ -3,6 +3,7 @@
  * Copyright (c) 2002-2014, Jouni Malinen <j@w1.fi>
  * Copyright (c) 2007, Johannes Berg <johannes@sipsolutions.net>
  * Copyright (c) 2009-2010, Atheros Communications
+ * Copyright (C) 2015 Freescale Semiconductor, Inc.
  *
  * This software may be distributed under the terms of the BSD license.
  * See README for more details.
@@ -394,6 +395,13 @@ static void mlme_event_disconnect(struct wpa_driver_nl80211_data *drv,
 			wpa_printf(MSG_DEBUG, "nl80211: Ignore disconnect "
 				   "event triggered during reassociation");
 			return;
+#ifdef BROADCOM_WIFI_VENDOR
+		} else if ((WLAN_REASON_DEAUTH_LEAVING == nla_get_u16(reason)) &&
+                !(drv->associated)) {
+            wpa_printf(MSG_WARNING, "nl80211: Ignore disconnect "
+                            "supplicant has already updated its state machine");
+            return;
+#endif
 		}
 		wpa_printf(MSG_WARNING, "nl80211: Was expecting local "
 			   "disconnect but got another disconnect "
diff --git a/wpa_supplicant/Android.mk b/wpa_supplicant/Android.mk
index 32e3c82..d3519bc 100644
--- a/wpa_supplicant/Android.mk
+++ b/wpa_supplicant/Android.mk
@@ -1577,6 +1577,7 @@ LOCAL_STATIC_LIBRARIES += libnl_2
 endif
 endif
 LOCAL_CFLAGS := $(L_CFLAGS)
+LOCAL_CFLAGS += -DBROADCOM_WIFI_VENDOR
 LOCAL_SRC_FILES := $(OBJS)
 LOCAL_C_INCLUDES := $(INCLUDES)
 include $(BUILD_EXECUTABLE)
-- 
1.8.0

